vtrace/envitools.py

```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
+                    self.syncRegsFromTrace()
+
+            except MemoryException as msg:
+                logger.warning("    \tError: %s: %s" % (repr(op), msg))
+                sys.stdin.readline()
+
+            except envi.SegmentationViolation as msg:
+                logger.warning("    \tError: %s: %s" % (repr(op), msg))
+                sys.stdin.readline()
+
+            except Exception as msg:
+                logger.warning("\t\tLockstep Error: %s" % msg, exc_info=1)
+                return
+            print('')
+
+    def syncRegsFromTrace(self):
+        rsnap = self.trace.getRegisterContext().getRegisterSnap()
+        self.emu.setRegisterSnap(rsnap)
+
+    def cmpRegs(self):
+        ctx = self.trace.getRegisterContext()
+        out = []
+        for rname, idx in ctx.getRegisterNameIndexes():
+            er = self.emu.getRegister(idx)
+            tr = self.trace.getRegisterByName(rname)
+            #tr = self.trace.getRegister(idx)
+            # debug registers aren't really used much anymore...
+            if er != tr and not rname.startswith('debug'):
+                out.append("%s (trace: 0x%.8x) (emulated: 0x%.8x)" % (rname, tr, er))
+
+        if len(out):
+            raise RegisterException("REGISTER MISMATCH: %s" % '    '.join(out))
 
-    # could use {get,set}MemorySnap if trace inherited from MemoryObject
-    for va, size, perms, fname in trace.getMemoryMaps():
-        try:
-            # So linux maps in a PROT_NONE page for efficient library sharing, so we have to take that into account
-            if (not perms & e_memory.MM_READ):
-                continue
-            if plat == 'linux' and fname in ['[vvar]']:
+        return True
+
+    def grabMemAddrs(self, op):
+        '''
+        Grab target addresses of opcode operands (before executing)
+        for comparison of memory after execution.
+
+        Also, compares the addresses calculated by Trace and EMU.
+        '''
+        self._tempaddrs = []
+        for oper in op.opers:
+            # check if we access memory # TODO: use read/write-logging instead
+            if not oper.isDeref():
                 continue
-            bytez = trace.readMemory(va, size)
-            emu.addMemoryMap(va, perms, fname, bytez)
-        except vtrace.PlatformException:
-            print('failed to map: 0x{:x} into emu'.format(va, size))
-            continue
 
-    rsnap = trace.getRegisterContext().getRegisterSnap()
-    emu.setRegisterSnap(rsnap)
+            taddr = oper.getOperAddr(op, self.trace) # use emu here?  it will likely fail the same either way? emu is faster?
+            taddr_emu = oper.getOperAddr(op, self.emu) # use emu here?  it will likely fail the same either way? emu is faster?
+            if taddr != taddr_emu:
+                raise TargetAddrCalcException('TARGET ADDRESS CALCULATED DIFFERENTLY: 0x%x: %r   (%r: emu: 0x%x,  trace: 0x%x)'\
+                        % (op, oper, taddr, taddr_emu))
```

Comment:
Uhh, you probably want `op.va` instead of op for the first param, since it looks like you're trying to repr an address?

Also, I'm not a fan of just dumping the raw oper. One thing that I just haven't gotten around to (since it's relatively low priority) is changing how opers repr themselves. Right now lot of the operands in like i386 have a `repr` function, but python's builtin stuff calls `__repr__`, so we end up just getting the raw python `<foo class @ anAddressThatIsMeaningless`, so that's something we've gotta fix.

TL;DR: you also probably want `oper.repr(op)` until we fix things.

Reply:
righto on the op.va!

actually, i want the `<foo class @.....` here, since i want to know what the class is that is a part of the problem.  Sib and Mem can look an awful lot alike.